File: src/SemanticTypeForgers/forger.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace SemanticTypeForgers
{
    public class Forger
    {
        public class AtomicBehavior<T>
        {
            private string name;

            public AtomicBehavior(string name)
            {
                this.name = name;
            }

            public T GetPrimitiveType(T v)
            {
                return v;
            }

            public bool Validate(object value)
            {
                return false;
            }

            private object PrimitiveStringToValue(string val)
            {
                string trimmed = val.Trim();
                string lower = trimmed.ToLower();

                if (string.IsNullOrEmpty(trimmed)) return val;
                if (val == "{}" || val == "[]") return 0;
                if (lower == "true") return true;
                if (lower == "false") return false;
                if (lower == "null" || lower == "undefined") return null;

                if (Regex.IsMatch(trimmed, @"^[+-]?(?:(?:\d+\.?\d*)|(?:\.\d+))(?:e[+-]?\d+)?$", RegexOptions.IgnoreCase))
                {
                    if (trimmed.Contains(".") || lower.Contains("e"))
                    {
                        if (double.TryParse(trimmed, out double d)) return d;
                    }
                    else
                    {
                        if (long.TryParse(trimmed, out long l)) return l;
                    }
                }

                return val;
            }

            private object Unwrap(object val, HashSet<int> seen, int depth)
            {
                if (depth > 100) throw new Exception("convertToPrimitive: recursive unwrap limit exceeded");
                if (val == null) return null;

                if (val is int || val is long || val is double || val is bool) return val;
                if (val is string s) return PrimitiveStringToValue(s);

                // Collections handling could be added here
                return null; // simplification
            }

            public object ConvertToPrimitive(object value)
            {
                try
                {
                    return Unwrap(value, new HashSet<int>(), 0);
                }
                catch
                {
                    return null;
                }
            }

            public T Forge(T v)
            {
                return v;
            }

            public T ProcessValue(object value)
            {
                if (Validate(value)) return (T)value;

                if (value is Dictionary<string, object> obj)
                {
                    double GetVal(string key)
                    {
                        if (obj.TryGetValue(key, out var v))
                        {
                            var p = ConvertToPrimitive(v);
                            if (p is IConvertible c) return c.ToDouble(null);
                        }
                        return 0;
                    }

                    double pPrice = GetVal("productPrice");
                    double pDiscount = GetVal("productDiscount");
                    double dPrice = GetVal("deliveryPrice");
                    double pFees = GetVal("paymentFees");

                    if (obj.ContainsKey("productPrice"))
                    {
                        double finalPrice = pPrice - pDiscount + dPrice + pFees;
                        return Forge((T)Convert.ChangeType(finalPrice, typeof(T)));
                    }
                }

                object primitiveVal = ConvertToPrimitive(value);
                return Forge((T)Convert.ChangeType(primitiveVal, typeof(T)));
            }
        }
    }
}

File: src/SemanticTypeForgers/forger.erl
-module(forger).
-export([get_primitive_type/1, validate/1, convert_to_primitive/1, forge/1, proccess_value/1]).

get_primitive_type(V) -> V.

validate(_) -> false.

trim(S) -> string:trim(S).
to_lower(S) -> string:lowercase(S).

primitive_string_to_value(Val) ->
    Trimmed = trim(Val),
    Lower = to_lower(Trimmed),
    if
        Trimmed == "" -> Val;
        Val == "{}"; Val == "[]" -> 0;
        Lower == "true" -> true;
        Lower == "false" -> false;
        Lower == "null"; Lower == "undefined" -> null;
        true ->
            case re:run(Trimmed, "^[+-]?(?:(?:\\d+\\.?\\d*)|(?:\\.\\d+))(?:e[+-]?\\d+)?$", [caseless]) of
                {match, _} ->
                    HasDot = string:chr(Trimmed, $.) > 0,
                    HasE = string:chr(Lower, $e) > 0,
                    if
                        HasDot orelse HasE ->
                            {F, _} = string:to_float(Trimmed ++ if HasDot -> ""; true -> ".0" end),
                            F;
                        true ->
                            {I, _} = string:to_integer(Trimmed),
                            I
                    end;
                nomatch -> Val
            end
    end.

unwrap(Val, _Seen, Depth) when Depth > 100 -> erlang:error(recursive_unwrap_limit_exceeded);
unwrap(Val, _Seen, _Depth) when Val == null -> null;
unwrap(Val, _Seen, _Depth) when is_boolean(Val); is_integer(Val); is_float(Val) -> Val;
unwrap(Val, _Seen, _Depth) when is_list(Val) ->
    case io_lib:printable_list(Val) of
        true -> primitive_string_to_value(Val);
        false ->
            if
                length(Val) == 0 -> null;
                true -> unwrap(hd(Val), _Seen, _Depth + 1)
            end
    end;
unwrap(Val, _Seen, Depth) when is_map(Val) ->
    if
        map_size(Val) == 0 -> 0;
        true ->
            [First | _] = maps:values(Val),
            unwrap(First, _Seen, Depth + 1)
    end;
unwrap(_, _, _) -> null.

convert_to_primitive(Value) ->
    try
        unwrap(Value, sets:new(), 0)
    catch
        _:_ -> null
    end.

forge(V) -> V.

proccess_value(Value) ->
    case validate(Value) of
        true -> Value;
        false ->
            if
                is_map(Value) ->
                    PPrice = convert_to_primitive(maps:get(<<"productPrice">>, Value, maps:get("productPrice", Value, null))),
                    PDiscount = convert_to_primitive(maps:get(<<"productDiscount">>, Value, maps:get("productDiscount", Value, 0))),
                    DPrice = convert_to_primitive(maps:get(<<"deliveryPrice">>, Value, maps:get("deliveryPrice", Value, 0))),
                    PFees = convert_to_primitive(maps:get(<<"paymentFees">>, Value, maps:get("paymentFees", Value, 0))),

                    if
                        is_integer(PPrice) orelse is_float(PPrice) ->
                            forge(PPrice - PDiscount + DPrice + PFees);
                        true ->
                            forge(convert_to_primitive(Value))
                    end;
                true ->
                    forge(convert_to_primitive(Value))
            end
    end.
